test/unit/test-loading-indicator.js

```diff
@@ -0,0 +1,190 @@
+      .value(IntersectionObserverStub);
+
+    service = new LoadingIndicatorImpl(ampdoc);
+
+    loaderService = {
+      initializeLoader: env.sandbox.spy(),
+    };
+    env.sandbox
+      .stub(Services, 'extensionsFor')
+      .returns({installExtensionForDoc: () => sync()});
+    env.sandbox
+      .stub(Services, 'loaderServiceForDoc')
+      .returns(sync(loaderService));
+
+    el = document.createElement('div');
+  });
+
+  class IntersectionObserverStub {
+    constructor(callback, options) {
+      if (options.root && failRoot) {
+        throw new Error('root is not allowed');
+      }
+      this.callback = callback;
+      this.options = options;
+      this.observed = [];
+      io = this;
+    }
+
+    observe(el) {
+      pushIfNotExist(this.observed, el);
+    }
+
+    unobserve(el) {
+      removeItem(this.observed, el);
+    }
+
+    record(record) {
+      this.callback([record]);
+    }
+  }
+
+  function sync(value) {
+    return {then: (callback) => callback(value)};
+  }
+
+  function getLoader() {
+    return el.querySelector('.i-amphtml-loading-container');
+  }
+
+  it('should use root when supported', () => {
+    failRoot = false;
+    service = new LoadingIndicatorImpl(ampdoc);
+    expect(io.options.root).to.equal(ampdoc.win.document);
+
+    service.track(el);
+    expect(io.observed).to.include(el);
+  });
+
+  it('should fail over when root is not supported', () => {
+    failRoot = true;
+    service = new LoadingIndicatorImpl(ampdoc);
+    expect(io.options.root).to.be.undefined;
+
+    service.track(el);
+    expect(io.observed).to.include(el);
+  });
+
+  it('should not create the loader on out-of-viewport', async () => {
+    service.track(el);
+    // No loader yet.
+    expect(getLoader()).to.not.exist;
+
+    // Observe out-of-viewport.
+    io.record({
+      target: el,
+      isIntersecting: false,
+      boundingClientRect: {width: 100, height: 100},
+    });
+    expect(getLoader()).to.not.exist;
+    expect(loaderService.initializeLoader).to.not.be.called;
+  });
+
+  it('should create the loader on the first in-viewport', async () => {
+    service.track(el);
+    // No loader yet.
+    expect(getLoader()).to.not.exist;
+
+    // Observe in-viewport.
+    io.record({
+      target: el,
+      isIntersecting: true,
+      boundingClientRect: {width: 100, height: 100},
+    });
+    expect(getLoader()).to.exist;
+    expect(loaderService.initializeLoader).to.be.calledOnce;
+  });
+
+  it('should NOT create the loader when size is too small', async () => {
+    service.track(el);
+    // No loader yet.
+    expect(getLoader()).to.not.exist;
+
+    // Observe in-viewport.
+    io.record({
+      target: el,
+      isIntersecting: true,
+      boundingClientRect: {width: 1, height: 1},
+    });
+    expect(getLoader()).to.not.exist;
+    expect(loaderService.initializeLoader).to.not.be.called;
+  });
+
+  it('should hide the loader on out-of-viewport', async () => {
```

Comment:
FMI: Is there a performance cost to leaving loaders on out-of-viewport items?

Reply:
Yes. They continue to animate. Eventually I hope that some of it will be manageable using `content-visibility`. But for now this has to be done manually. And it's the old behavior as well.